File: src/bcs/sdks/java/src/main/java/com/avernet/bcs/provider/sdk/example/EchoServer.java
package com.avernet.bcs.provider.sdk.example;

import com.avernet.bcs.provider.sdk.ProviderGrpcServer;
import com.avernet.bcs.provider.sdk.ProviderService;

/** Executable example of a Provider SDK subclass. */
public final class EchoServer {
    private EchoServer() {
    }

    public static void main(String[] args) throws Exception {
        Arguments arguments = Arguments.parse(args);
        final ProviderGrpcServer server = new ProviderGrpcServer(
                new EchoProvider(), arguments.host, arguments.port);
        server.start();
        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            try {
                server.close();
            } catch (InterruptedException error) {
                Thread.currentThread().interrupt();
            }
        }));
        System.out.println(
                "Java Provider demo listening on " + arguments.host + ":" + server.getPort());
        server.blockUntilShutdown();
    }

    private static final class EchoProvider extends ProviderService {
        @Override
        public String implementation() {
            return "java";
        }

        @Override
        public String invoke(String message) {
            return "java: " + message;
        }
    }

    private static final class Arguments {
        private final String host;
        private final int port;

        private Arguments(String host, int port) {
            this.host = host;
            this.port = port;
        }

        private static Arguments parse(String[] args) {
            String host = "127.0.0.1";
            int port = 50052;
            for (int index = 0; index < args.length; index++) {
                String option = args[index];
                if ("--host".equals(option)) {
                    host = requireValue(args, ++index, option);
                } else if ("--port".equals(option)) {
                    port = Integer.parseInt(requireValue(args, ++index, option));
                } else {
                    throw new IllegalArgumentException("unknown argument: " + option);
                }
            }
            return new Arguments(host, port);
        }

        private static String requireValue(String[] args, int index, String option) {
            if (index >= args.length) {
                throw new IllegalArgumentException("missing value for " + option);
            }
            return args[index];
        }
    }
}

File: src/bcs/sdks/python/examples/echo_server.py
import argparse
import asyncio

from bcs_provider_sdk import ProviderServer, ProviderService


class EchoProvider(ProviderService):
    @property
    def implementation(self) -> str:
        return "python"

    async def invoke(self, message: str) -> str:
        return f"python: {message}"


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description="Run the Python Provider SDK demo")
    parser.add_argument("--host", default="127.0.0.1")
    parser.add_argument("--port", default=50051, type=int)
    return parser.parse_args()


async def run(host: str, port: int) -> None:
    server = ProviderServer(EchoProvider(), host=host, port=port)
    await server.start()
    print(f"Python Provider demo listening on {host}:{server.bound_port}", flush=True)
    try:
        await server.wait_for_termination()
    finally:
        await server.stop()


if __name__ == "__main__":
    args = parse_args()
    try:
        asyncio.run(run(args.host, args.port))
    except KeyboardInterrupt:
        pass
